text/3753-rest-pattern-matches-non-struct.md

```diff
@@ -0,0 +1,138 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+(The following text is appended to the section on [struct
+patterns](https://doc.rust-lang.org/stable/reference/patterns.html#struct-patterns)
+in the reference.)
+
+As a special case, if a struct pattern contains only `..` as the fields, and the
+path refers to a type (as opposed to an enum variant), then the pattern is an
+irrefutable pattern that matches against any value of that type. This applies
+even if that type is not a struct type (e.g., it might be an enum type, or it
+might be a type alias, etc.)
+
+For example, the pattern `foo::Bar::<Baz> { .. }` can match any value of the
+type `foo::Bar<Baz>` (even if this type is not a struct type), or match a value
+of the type `foo<Baz>` that contains the enum variant `Bar`.
+
+Formally, this special case applies to the following syntax: *PathInExpression*
+`{` *StructPatternEtCetera* `}`
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds a complication to Rust. 
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* This special case is only applied to `Foo { .. }`, and not `Foo(..)`.
+  * This is because, currently, the pattern `Foo(..)` can only match a tuple
+  struct whose fields are all public, so the `Foo(..)` pattern does not pose a
+  semver hazard.
+  * On the other hand, the pattern `Foo { .. }` currently matches any struct
+  named `Foo`, including tuple structs, and including type aliases that refer to
+  structs. This makes the meaning of the pattern `Foo { .. }` already similar to
+  "match any type named `Foo`".
+* As an alternative, we could deprecate the pattern `Foo { .. }` (either in all
+  cases, or only in cases where `Foo` has no public fields). We could then
+  potentially remove this pattern from the language in a future edition.
+  * This unfortunately doesn't fix the semver hazard, due to code in older
+    editions existing. Additionally, this might be an edge case that macros
+    would have to deal with.
+* As an alternative, we could have an attribute that marks a type as completely
+  opaque, and therefore making it not able to be matched with the pattern `Foo {
+  .. }`.
+  * Most users are likely to forget to apply this attribute. We could change the
+    default over an edition, making a struct with no public fields implicitly
+    opaque, but this special case seems rather weird and confusing.
+
+# Prior art
+[prior-art]: #prior-art
+
+`cargo-semver-checks` has [a
+lint](https://github.com/obi1kenobi/cargo-semver-checks/issues/954) that checks
+specifically if a struct containing only private fields is changed to a
+different type.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* How much code in the wild currently uses patterns like `Foo { .. }` ?
+* What was the original reason that the pattern `Foo { .. }` matches all
+  structs, and not just tuple-like structs?
```

Comment:
This is because all structures are just that, structures. 

Unit structures are just a sugar for a struct with a constant: 
```rust
struct Unit;
// <=>
struct Unit {}
const Unit: Unit = Unit {};
```

Similarly tuple structs are just a sugar for a struct with friends named as integers, a function and a pattern (these are but expressible in surface level rust, but still):
```rust
struct Tuple(u8);
// <=>
struct Tuple { 0: u8 }
fn Tuple(_0: u8) -> Tuple { Tuple { 0: _0 } }
pattern Tuple(_0) = Tuple { 0: _0 }
```

So, for most intents and purposes all kinds of structs are the same.

(also this probably has a typo, I assume it should have said "not just named structs"?)

Reply:
After reading RFC 1506, I'm still rather confused on the motivation. It seems that the reason that a pattern like `Tuple { 0: x }` has the same meaning as `Tuple(x)` is for... macros? Am I missing something?